### DISCUSSION TITLE
Get a list of tokens from a tree?

### DISCUSSION THREAD

**Discussion by chiefchewie** [tree-sitter/tree-sitter#2001] [Q&A] [answered]
I'm trying get a list of tokens from a source code file. That is, break down the text into tokens and have some information regarding each token, i.e. this token is an "identifier", "string", or "equality operator" and so on. As I far as I understand it, this is what a lexer does, but from [this discussion](https://github.com/tree-sitter/tree-sitter/discussions/1036) it seems that lexing and parsing happen at the same time. So it seems like I can't just ask tree-sitter to only do the lexing and give me the tokens.

Now I am wondering if I can somehow get tokens and their context back using a tree parsed by tree-sitter?

**Comment by ahlinc** [comment, marked as answer]
@chiefchewie It's possible by traversing a tree in a depth first fashion and append to a list every node that is a leaf node.

**Reply by chiefchewie** [reply]
Thanks.
